### DISCUSSION TITLE
Shortcomings in SAX interface

### DISCUSSION THREAD

**Discussion by nlohmann** [nlohmann/json#2284] [Q&A]
In the light of #2275, I am currently investigating how hard it is to use the SAX interface internally in the serializers.

For this, I implemented a small member function `visit` that traverses the JSON value and fires the respective SAX events:

```cpp
template<typename SAX>
JSON_HEDLEY_NON_NULL(2)
void visit(SAX* sax) const
{
    (void)detail::is_sax_static_asserts<SAX, basic_json> {};

    switch (type())
    {
        case value_t::null:
        {
            sax->null();
            break;
        }

        case value_t::boolean:
        {
            sax->boolean(m_value.boolean);
            break;
        }

        case value_t::number_integer:
        {
            sax->number_integer(m_value.number_integer);
            break;
        }

        case value_t::number_unsigned:
        {
            sax->number_unsigned(m_value.number_unsigned);
            break;
        }

        case value_t::number_float:
        {
            sax->number_float(m_value.number_float, "");
            break;
        }

        case value_t::string:
        {
            sax->string(*m_value.string);
            break;
        }

        case value_t::binary:
        {
            sax->binary(*m_value.binary);
            break;
        }

        case value_t::object:
        {
            sax->start_object(m_value.object->size());
            for (const auto& entry : *m_value.object)
            {
                sax->key(entry.first);
                entry.second.visit(sax);
            }
            sax->end_object();
            break;
        }

        case value_t::array:
        {
            sax->start_array(m_value.array->size());
            for (const auto& entry : *m_value.array)
            {
                entry.visit(sax);
            }
            sax->end_array();
            break;
        }
    }
}
```

The idea is then to implement the serializer via the SAX interface. However, I found (at least) the following issues:

- The `number_float` event takes the original string of the parsed floating-point number as second parameter. This string is present when the SAX interface is used by the parser, but not if I access a floating-point number stored in a `basic_json` object later. Passing the empty string would need some kind of contract that it means "not available, stringify the value yourself if needed".
- The `string`, `binary` and `key` events assume a non-const reference - at least this is how they are defined in the `basic_json::json_sax_t` interface. The reason was that when the value is created by the parser, then it would be allowed to move it to the consumer of the SAX events. However, I would like to have a `const` visit function, because serialization should be `const`. The function above hence does not work with subclasses of `basic_json::json_sax_t`.
- There is no event for discarded values.

Any ideas how to proceed here?

**Comment by flyx** [comment]
Would you consider using/creating a different interface that more closely resembles an event stream (while keeping `json_sax_t`)? Like I mentioned, SAX has the problem that control cannot be handed over to another object during parsing, and having a more flexible interface would benefit both loading and serializing. I could try to sketch some code if you're interested.

**Reply by nlohmann** [reply]
Using a different interface is a pragmatic idea to try the approach. As I don't fully understand what you mean with control being handled over - yes, I am interested in some code!

**Reply by flyx** [reply]
Okay so this will be mostly a translation of what I do in NimYAML to pseudocode C++ which will not necessarily compile directly but shows the general idea.

I propose a polling API, rather than a signalling API. A very basic API could look like this:

```c++
enum class dom_event {
    null, boolean, number_integer, number_unsigned, number_float, string, binary,
    start_object, key, end_object, start_array, end_array
};

// types assumed to be from basic_json
class event_stream {
  public:
    virtual dom_event next() =0;

    virtual bool get_boolean() =0;
    virtual number_integer_t get_integer() =0;
    virtual number_unsigned_t get_unsigned() =0;
    virtual number_float_t get_float() =0;
    virtual const binary_t &get_binary() =0;
    virtual const string_t& get_string() =0;
};
```

`event_stream` would (semantically, not in the C++ sense) be an iterator over a stream of events. `next()` returns the next element. Since different events carry different payloads, getters for each type of payload are provided (`key` events would use `get_string`).

The most obvious drawback is that this does not establish the relation between event and payload, like the SAX API does (i.e. the code does not make it obvious what kind of payload you can query for the current event). But the added flexibility can be very useful. This interface can be provided by

 * The parser, transforming input into an event stream
 * `basic_json`, transforming its content into an event stream
 * Native objects, transforming their content into an event stream (this would be the most difficult, and it's not obvious how to autogenerate it)

And it can be consumed by

 * The renderer, transforming an event stream into a JSON representation
 * `basic_json`, loading the event stream into a tree
 * Native objects, loading the stream into their native fields / structure.

Compared to the SAX interface, some information is missing (number of items in objects/arrays, float representation), but can easily be added by the implementations. For example, the parser implementation of `event_stream` could provide an additional getter for the float representation (or, in fact, any representation, which would enable us to see which escape sequences were employed in a string), and the `basic_json` and native object implementations could provides getters for the number of items in objects/arrays. Or those getters can be available in the base interface along with a way to query the interface for its capabilities of providing those items.

Now to show what I mean by handing over control, here's some pseudoish code showing how we could consume the stream to build a `basic_json` object:

```c++
void from_json_stream(event_stream& s, json& j) {
    switch (auto evt = s.next(); evt) {
        case dom_event::null: j = nullptr; return;
        case dom_event::boolean: j = s.get_boolean(); return;
        case dom_event::number_integer: j = s.get_integer(); return;
        case dom_event::number_unsigned: j = s.get_unsigned(); return;
        case dom_event::number_float: j = s.get_float(); return;
        case dom_event::string: j = s.get_string(); return;
        case dom_event::binary: j = s.get_binary(); return;
        case dom_event::start_object: {
            j = {};
            for (evt = s.next(); evt != dom_event::end_object; evt = s.next()) {
                assert(evt == dom_event::key); // must be true for well-formed stream
                auto key = s.get_string();
                json value;
                // hand over control to loader of the child value here
                from_json_stream(s, value);
                j[key] = value;
            }
            return;
        }
        case dom_event::start_array: {
            j = {};
            for (evt = s.next(); evt != dom_event::end_array; evt = s.next()) {
                json item;
                // same here
                from_json_stream(s, item);
                j.push_back(item);
            }
            return;
        }
        default: assert(false); // must not happen in well-formed stream
    }
}
```

And loaders for native types could work in a similar way: struct types will expect a `dom_event::start_object`, then map the keys to the fields and for each field, hand over the stream to the loader of the field value's type.

Implementing the API would be a bit more involved, here's some horrible code that tries to do it for `basic_json` with no claim of correctness whatsoever, let alone safety:

```c++
// implied to be a nested class in basic_json
class object_stream: public event_stream {
    union {
        bool bVal;
        number_integer_t i;
        number_unsigned_t u;
        number_float_t f;
        const binary_t *b;
        const string_t *s;
    } value;

    class level {
        bool initial;
        dom_event kind;
        object_t::const_iterator o_iter, o_iter_end;
        array_t::const_iterator a_iter, a_iter_end;
      public:
        level(const basic_json& j, object_stream &s): initial(true) {
            switch (j.type()) {
                case value_t::null:
                    kind = dom_event::null;
                    break;
                case value_t::boolean:
                    kind = dom_event::boolean;
                    s.value.b = j.m_value.boolean;
                    break;
                case value_t::number_integer:
                    kind = dom_event::number_integer;
                    s.value.i = j.m_value.number_integer;
                    break;
                case value_t::number_unsigned:
                    kind = dom_event::number_unsigned:
                    s.value.u = j.m_value.number_unsigned;
                    break;
                case value_t::number_float:
                    kind = dom_event::number_float;
                    s.value.f = j.m_value.number_float;
                    break;
                case value_t::string:
                    kind = dom_event::string;
                    s.value.s = j.m_value.string;
                    break;
                case value_t::binary:
                    kind = dom_event::binary;
                    s.value.b = j.m_value.binary;
                    break;
                case value_t::object:
                    kind = dom_event::start_object;
                    o_iter = j.m_value.object->begin();
                    o_iter_end = j.m_value.object->end();
                    break;
                case value_t::array:
                    kind = dom_event::start_array;
                    a_iter = j.m_value.array->begin();
                    a_iter_end = j.m_value.array->end();
                    break;
            }
        }
        // return false if level is finished
        bool advance(object_stream &s, dom_event &e) {
            if (initial) {
                e = kind;
                return kind == dom_event::start_object || kind == dom_event::start_array;
            } else switch (kind) {
                case dom_event::start_object:
                    if (o_iter == o_iter_end) {
                        e = dom_event::end_object;
                        return false;
                    }
                    s.value.s = o_iter->first;
                    e = kind = dom_event::key;
                    return true;
                case dom_event::key:
                    level next(o_iter->second);
                    kind = dom_event::start_object;
                    ++o_iter;
                    s.levels.push_back(next); // may invalidate this ptr
                    return s.levels.back().advance(s, e);
                case dom_event::start_array:
                    if (a_iter == a_iter_end) {
                        e = dom_event::end_array;
                        return false;
                    }
                    level next(*a_iter);
                    kind = dom_event::start_array;
                    ++a_iter;
                    s.levels.push_back(next);
                    return s.levels.back().advance(s, e);
                default:
                    // may never happen
            }
        }
    };
    std::vector<level> levels;
  public:
    event_stream(const basic_json& j) {
        levels.emplace_back(j);
    }

    dom_event next() {
        dom_event e;
        if (!levels.back().advance(*this, e)) {
            levels.pop_back();
        }
        return e;
    }

    bool get_boolean() {
        return value.b;
    }
    number_integer_t get_integer() {
        return value.i;
    }
    number_unsigned_t get_unsigned() {
        return value.u;
    }
    number_float_t get_float() {
        return value.f;
    }
    const binary_t& get_binary() {
        return *value.b;
    }
    const string_t& get_string() {
        return *value.s;
    }
};
```

This code can be generalized to work with native objects too, if we transform the `level` type into an abstract class and generate a child class for each type that is to be serialized.

(Wow this got to be a lot longer than expected)

**Reply by nlohmann** [reply]
Thanks, I'll have a look!

**Comment by Ericson2314** [comment]
Maybe check out this PR https://github.com/NixOS/nix/pull/3893? We were all excited to switch everything over to just use nlohmann/json, but we noticed this would cause O(n) memory usage/delay. I do think the old destructer-based interface given one a degree of structure while maintaining constant-space usage is pretty neat. Perhaps that would solve some of the issues in this thread?

**Comment by Ericson2314** [comment]
@nlohmann 

Any more progress on this? Here are some ideas for the problem you mentioned before:

> Passing the empty string would need some kind of contract that it means "not available, stringify the value yourself if needed".

Maybe just use `std::optional`? I am not sure what the tolerance for breaking changes is.

> The string, binary and key events assume a non-const reference...

I would just do what Rust does in https://doc.rust-lang.org/std/borrow/enum.Cow.html, this "maybe owned" thing for the string case is tantamount to `std::variant<std::string, std::string_view>`, and likewise for the other cases. It should be easy for the consumer to handle both cases, and this gives the producer flexibility.

(BTW see https://oribenshir.github.io/afternoon_rusting/blog/copy-on-write for why this is quite unrelated to traditional C++ assignment operator COW, despite sharing the name.)

> There is no event for discarded values.

Not sure what you mean by this.

**Reply by Ericson2314** [reply]
If on the other hand you want the parse to statically choose whether it wants `const` or not, `std::variant` might be useful? You could have a `std::variant` of all the the events, and a unified callback taking the `std::variant`. However, `std::visit` works by value or by (`const`, or non-`const`, l or r) reference, and so you can expose variations of the parsing function fairly easily.
